src/java/main/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -866,6 +866,9 @@ public void setServerCnxnFactory(ServerCnxnFactory factory) {
     }
 
     public ServerCnxnFactory getServerCnxnFactory() {
+        if (secureServerCnxnFactory != null) {
+            return secureServerCnxnFactory;
+        }
         return serverCnxnFactory;
     }
 
```

Comment:
Would it be better to, instead of mix `secureServerCnxnFactory` in `getServerCnxnFactory`, add a separate method `getSecureServerCnxnFactory`? This also maps well with existing set method `setSecureServerCnxnFactory`. Caller now has to explicitly call both, which is more work, but makes semantics more clear.

Reply:
i don't feel strongly about it, but i like @hanm 's idea.

Reply:
I will look into it. Original issue was on the caller side which didn't have the logic to probe which ServerCnxnFactory is available, so I probably have to make it more clever.

Though I'm not entirely convinced that it should be the caller's responsibility to deal with the problem. He just want something that implements the interface. Why should make this their problem?

Reply:
Makes sense.